File: app/src/main/java/org/blitzortung/android/app/components/VersionComponent.java
package org.blitzortung.android.app.components;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

public class VersionComponent {

    private State state;
    private String versionName;

    public enum State {
        FIRST_RUN, FIRST_RUN_AFTER_UPDATE, NO_UPDATE
    }

    private static final String CONFIGURED_VERSION_CODE = "configured_version_code";

    private int configuredVersionCode;
    private int currentVersionCode;

    public VersionComponent(final Context context) {
        updatePackageInfo(context);
        updateVersionStatus(context);
    }

    private void updateVersionStatus(Context context) {
        final SharedPreferences preferences = context.getSharedPreferences(context.getPackageName(), Context.MODE_PRIVATE);
        configuredVersionCode = preferences.getInt(CONFIGURED_VERSION_CODE, -1);

        if (configuredVersionCode == -1) {
            state = State.FIRST_RUN;
        } else {
            if (configuredVersionCode < currentVersionCode) {
                state = State.FIRST_RUN_AFTER_UPDATE;
            } else {
                state = State.NO_UPDATE;
            }
        }
    }

    private void updatePackageInfo(final Context context) {
        final PackageInfo pInfo;
        try {
            pInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
        } catch (PackageManager.NameNotFoundException e) {
            throw new IllegalStateException(e);
        }

        currentVersionCode = pInfo.versionCode;
        versionName = pInfo.versionName;
    }

    public State getState() {
        return state;
    }

    public int getConfiguredVersionCode() {
        return configuredVersionCode;
    }

    public int getVersionCode() {
        return currentVersionCode;
    }

    public String getVersionName() {
        return versionName;
    }
}

File: app/src/main/java/org/blitzortung/android/app/components/VersionComponent.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.app.components

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.core.content.edit
import javax.inject.Inject
import org.blitzortung.android.app.BuildConfig
import org.blitzortung.android.app.Main

class VersionComponent
    @Inject
    constructor(
        context: Context,
        buildVersion: BuildVersion,
    ) {
        var state: State? = null
            private set

        var configuredVersionCode: Int = 0
            private set

        var configuredMajorVersion: Int = 0
            private set

        var configuredMinorVersion: Int = 0
            private set

        private val preferences: SharedPreferences

        init {
            val packageName = context.packageName
            preferences = context.getSharedPreferences(packageName, Context.MODE_PRIVATE)

            configuredVersionCode = getAndStoreVersion(CONFIGURED_VERSION_CODE, buildVersion.versionCode)
            configuredMajorVersion = getAndStoreVersion(CONFIGURED_MAJOR_VERSION, buildVersion.majorVersion)
            configuredMinorVersion = getAndStoreVersion(CONFIGURED_MINOR_VERSION, buildVersion.minorVersion)

            state =
                when {
                    configuredVersionCode == -1 -> State.FIRST_RUN
                    configuredMajorVersion != buildVersion.majorVersion ||
                        configuredMinorVersion != buildVersion.minorVersion -> State.FIRST_RUN_AFTER_UPDATE

                    else -> State.NO_UPDATE
                }

            Log.d(
                Main.LOG_TAG,
                "updateVersionStatus() name=$packageName, state=$state, versionCode=$configuredVersionCode/${buildVersion.versionCode}, major=$configuredMajorVersion/${buildVersion.majorVersion} minor=$configuredMinorVersion/${buildVersion.minorVersion}",
            )
        }

        private fun getAndStoreVersion(
            prefsKey: String,
            newValue: Int,
        ): Int {
            val knownValue = preferences.getInt(prefsKey, -1)
            preferences.edit { putInt(prefsKey, newValue) }
            return knownValue
        }

        enum class State {
            FIRST_RUN,
            FIRST_RUN_AFTER_UPDATE,
            NO_UPDATE,
        }

        companion object {
            internal const val CONFIGURED_VERSION_CODE = "configured_version_code"
            internal const val CONFIGURED_MAJOR_VERSION = "configured_major_version"
            internal const val CONFIGURED_MINOR_VERSION = "configured_minor_version"
        }
    }

class BuildVersion
    @Inject
    constructor() {
        val versionCode = BuildConfig.VERSION_CODE
        var versionName = BuildConfig.VERSION_NAME
        val majorVersion: Int
        val minorVersion: Int
        val patchVersion: Int

        init {
            val versionComponents = versionName.split(".")
            majorVersion = versionComponents[0].toInt()
            minorVersion = versionComponents[1].split('-')[0].toInt()
            patchVersion = if (versionComponents.size > 2) versionComponents[2].toInt() else -1
        }
    }
